### PR TITLE
[Discussion]Create show_result.sh

### PR DESCRIPTION
**Pull request by kamo-naoyuki**
#786

This file is an enhancement of `utils/get_sys_info.sh`.

- Usage:
```bash
cd egs/an4/asr1/
utils/show_result.sh
```

```markdown
<!-- Generated by ./show_result.sh -->
# RESULTS
## Environments
- date: `Tue Jun 18 18:22:28 JST 2019`
- system information: `Linux  <hostname> 3.10.0-957.1.3.el7.x86_64 #1 SMP Thu Nov 29 14:49:43 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`
- python version: `3.7.2 (default, Dec 29 2018, 06:19:36)  [GCC 7.3.0]`
- espnet version: `espnet 0.3.1`
- chainer version: `chainer 5.0.0`
- pytorch version: `pytorch 1.0.0`
- Git hash: `88c81722113bb83a20128c38eceeb951c2d7964e`
  - Commit date: `Sat May 25 06:55:17 2019 -0400`

## train_nodev_chainer_blstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150
### CER

|dataset|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|
|decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5|130|2565|81.2|7.4|11.3|4.2|23.0|74.6|
|decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5|100|1915|76.1|9.0|14.9|2.1|26.0|82.0|

## train_nodev_pytorch_blstmp_e4_subsample1_2_2_1_1_unit320_proj320_d1_unit300_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150
### CER

|dataset|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|
|decode_test_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5|130|2565|91.2|3.6|5.2|0.7|9.5|56.2|
|decode_train_dev_beam20_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.5|100|1915|84.4|5.4|10.1|1.5|17.1|70.0|

## train_nodev_pytorch_train_mtlalpha1.0
### CER

|dataset|Wrd|Corr|Sub|Del|Ins|Err|S.Err|
|---|---|---|---|---|---|---|---|
|decode_test_decode_ctcweight1.0_lm_word100|130|2565|92.9|3.4|3.7|1.1|8.2|52.3|
|decode_train_dev_decode_ctcweight1.0_lm_word100|100|1915|86.1|6.7|7.2|1.8|15.7|69.0|

```

### CODE REVIEW

**Comment by kamo-naoyuki** [comment]
I worry about uname -a. It may be not better to reveal the hostname for security reason.

**Comment by sw005320** [comment]
Oh, this is a very beautiful way!
You can replace my ugly `utils/get_sys_info.sh` with this.
Could you also modify `CCONTRIBUTING.md` accordingly?

Also, sure I agree with you that we should avoid having a hostname.
Could you implement it?

**Comment by ShigekiKarita** [comment]
Awesome. I have been always making some scripts like this when I wrote papers.

**Comment by kamo-naoyuki** [comment]
I think the training time and machine info, especially what GPU device was used, are also important.
 `uname -a` shows not the training machine, but actually the execution host of this script.

It is possible to get them from the `expdir` if `asr_train.py` saves such information, e.g.

```
exp/train/info/machine_info
exp/train/info/execution_time
```

(As `sacred` generates such files automatically,  I proposed it at the first time.
However, the unintelligibility from the too rich system is not acceptable comparing with sacred's benefit.
)

Currenly `exp/train/log` is available, though I don't know which module generates it, so training time already can be shown, but I think it is good to redesign of this log file.
Do you have any other useful information during training?

**Comment by ShigekiKarita** [comment]
On log files, maybe we should discuss in a new issue or #820

**Comment by sw005320** [comment]
> Do you have any other useful information during training?

- Pointer to the cmvn file (or we can link it under experiment directory), but not here?
- CUDA and CuDNN versions
- Number of GPUs
- Maximum GPU memory (I think this is difficult).

**Comment by kamo-naoyuki** [comment]
I decided to finish this PR and I'll modify this tool in another thread. 
Maybe, it is good to call show_result.sh at the last of run.sh and write it in exp/RESULT.md.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -24,16 +24,15 @@ If you are interested in contributing to ESPnet, your contributions will fall in
...
-	   You'll get environmental information in a markdown format.
+	 - execute `show_result.sh` at a recipe main directory (where `run.sh` is located), as follows. 
+	   You'll get environmental information and the evaluation result of each experiments in a markdown format.
 	   ```
-	   $ get_sys_info.sh
+	   $ show_result.sh
 	   ```
 	 - execute `pack_model.sh` at a recipe main directory as follows. You'll get model information and results in a markdown format
 	   ```
 	   $ pack_model.sh --lm <language model> --results "<dev result.txt> <test.result.txt>" <tr_conf> <dec_conf> <cmvn> <e2e>
```
Comment (sw005320) [inline CONTRIBUTING.md]: you could remove `--results "<dev result.txt> <test.result.txt>" `

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -24,16 +24,15 @@ If you are interested in contributing to ESPnet, your contributions will fall in
...
    To make a report for `RESULTS.md`
-	 - execute `get_sys_info.sh` at a recipe main directory (where `run.sh` is located), as follows. 
-	   You'll get environmental information in a markdown format.
+	 - execute `show_result.sh` at a recipe main directory (where `run.sh` is located), as follows. 
+	   You'll get environmental information and the evaluation result of each experiments in a markdown format.
 	   ```
-	   $ get_sys_info.sh
+	   $ show_result.sh
 	   ```
 	 - execute `pack_model.sh` at a recipe main directory as follows. You'll get model information and results in a markdown format
```
Comment (sw005320) [inline CONTRIBUTING.md]: You'll get model information and results in a markdown format
->
You'll get model information in a markdown format

**Comment by sw005320** [comment]
> Maybe, it is good to call show_result.sh at the last of run.sh and write it in exp/RESULT.md.

Nice idea.
My only concern is that this will write a lot of duplicated results and some unwanted results. Of course, people can edit it by themselves though.

**Comment by kamo-naoyuki** [comment]
> My only concern is that this will write a lot of duplicated results and some unwanted results. Of course, people can edit it by themselves though.

I thought the same as you too.
Wait a moment. I'll support only this in this PR.

**Comment by kamo-naoyuki** [comment]
By the way, well, I think markdown is useful for viewing on github, but at local machine environment, it is a bit troublesome.
Maybe, people would search a light tool for viewing markdown if we insert this tool in run.sh, but I don't know better choice for now. An editor such as Atom or a markdown extension(https://addons.mozilla.org/en-US/firefox/addon/markdown-viewer-chrome/) for browser looks useful.

**Comment by kamo-naoyuki** [comment]
O.K. I changed to use $(find ${exp} --mindepth 0 --maxdepth 1 -type d) instead of $(ls ${exp}/*).
Now both usages are supported.

```
show_result.sh  # Show all results
show_result.sh  exp/train_foobar # Show result only  exp/train_foobar
```

**Comment by sw005320** [comment]
> By the way, well, I think markdown is useful for viewing on github, but at local machine environment, it is a bit troublesome.

I'm using the emacs markdown extension, but this is not the perfect solution, and I'm checking it by pasting it somewhere.

**Pull request merged**